src/self_test.rs

```diff
@@ -0,0 +1,153 @@
+use std::process::Output;
+
+use tokio::process::Command;
+use which::which;
+
+#[non_exhaustive]
+#[derive(thiserror::Error, Debug, strum::IntoStaticStr)]
+pub enum SelfTestError {
+    #[error("Shell `{shell}` failed self-test with command `{command}`, stderr:\n{}", String::from_utf8_lossy(&output.stderr))]
+    ShellFailed {
+        shell: Shell,
+        command: String,
+        output: Output,
+    },
+    /// Failed to execute command
+    #[error("Failed to execute command `{command}`",
+        command = .command,
+    )]
+    Command {
+        shell: Shell,
+        command: String,
+        #[source]
+        error: std::io::Error,
+    },
+}
+
+#[cfg(feature = "diagnostics")]
+impl crate::diagnostics::ErrorDiagnostic for SelfTestError {
+    fn diagnostic(&self) -> String {
+        let static_str: &'static str = (self).into();
+        let context = match self {
+            Self::ShellFailed { shell, .. } => vec![shell.to_string()],
+            Self::Command { shell, .. } => vec![shell.to_string()],
+        };
+        return format!(
+            "{}({})",
```

Comment:
Do we want a space between these two?

```suggestion
            "{} ({})",
```

Reply:
In this case, nope! This is how we do it elsewhere.